src/Data/README.md

```diff
@@ -0,0 +1,73 @@
+# Datatypes for ILIAS
+
+This service should contain standard datatypes for ILIAS that are used in many
+locations in the system and do not belong to a certain service.
+
+## Result
+
+A result encapsulates a value or an error and simplifies the handling of those.
+
+```php
+<?php
+require_once(__DIR__."/Factory.php");
+
+$f = new Data/Factory;
+
+// Build a value that is ok.
+$pi = $f->ok(3.1415);
+
+// Value is ok and thus no error.
+assert($pi->isOK());
+assert(!$pi->isError());
+
+// Do some transformation with the value.
+$r = 10;
+$A = $f->map(function($pi) use ($r) { return 2 * $pi * $r; });
+
+// Still ok and no error.
+assert($pi->isOk());
+assert(!$pi->isError());
+
+// Retrieve the contained value.
+$A_value = $A->value();
+assert($A_value == 2 * 3.1415 * 10);
+
+// No error contained...
+$raised = false;
+try {
+	$A->error();
+	assert(false); // Won't happen, error raises.
+}
+catch(\LogicException $e) {
+	$raised = true;
+}
+assert($raised);
+
+
```

Comment:
`Your example is slightly wrong, the first function should return $f->ok($value_of_a * 20).`

Oke, along with the answer to one of the other question along with some already fainted knowledge about monads I got that.

`Basically because I had to explain Promises/monads then. I didn't feel powerful enough to do that in a Result-README.`

I am afraid, you will have to do that somewhere. My mistake above hints that. Your answer above helped me a lot and could be sufficient as explanation.